lineax/_operator.py

```diff
@@ -906,6 +906,78 @@ def out_structure(self):
         return self.operator.out_structure()
 
 
+class WoodburyLinearOperator(AbstractLinearOperator, strict=True):
+    """As [`lineax.MatrixLinearOperator`][], but for specifically a matrix
+    with A + U C V structure, such that the Woodbury identity can be used"""
+
+    A: AbstractLinearOperator
+    C: Inexact[Array, " k k"]
+    U: Inexact[Array, " n k"]
+    V: Inexact[Array, " k n"]
+    UCV: Array
+    tags: frozenset[object] = eqx.field(static=True)
+
+    def __init__(
+        self,
+        A: AbstractLinearOperator,
+        C: Inexact[Array, " k k"],
+        U: Inexact[Array, " n k"],
+        V: Inexact[Array, " k n"],
+        tags: Union[object, frozenset[object]] = (),
+    ):
+        """**Arguments:**
+
+        Matrix of form A + U C V, such that the inverse can be computed
+        using Woodbury matrix identity
+
+        - `A`: Linear operator, in/out shape (n,n)
+        - `C`: A rank-two JAX array. Shape (k,k)
+        - `U`: A rank-two JAX array. Shape (n,k)
+        - `V`: A rank-two JAX array. Shape (k,n)
+
+        """
+        self.A = A
+        self.C = inexact_asarray(C)
+        self.U = inexact_asarray(U)
+        self.V = inexact_asarray(V)
+        (N, M) = self.A.in_structure(), self.A.out_structure()
+        if N != M:
+            raise ValueError(f"expecting square operator for A, got {N} by {M}")
+        (K, L) = self.C.shape
+        if K != L:
+            raise ValueError(f"expecting square operator for C, got {K} by {L}")
+        N = N.shape[0]
```

Comment:
`N` is an arbitrary PyTree, it doesn't necessarily have a `.shape` attribute.

Reply:
Ok, This is mostly my ignorance of how lineax is using PyTrees as linear operators.

I am not sure how allowing arbitrary PyTrees for A meshes with the requirement of a Woodbury structure. In the context of the Woodbury matrix identity, I would think A would have to have a matrix representation of an n by n matrix. For a PyTree representation then would C, U and V also need to be PyTrees such that each leaf of the tree can be made to have Woodbury structure?

Reply:
Just thought I'd response on this since it has been a bit - thanks for the detailed response, it makes sense, I am (slowly) working on changes that would make the Woodbury implementation pytree compatible. 

I think the size checking is easy enough if I have understood you correctly here (PyTree-of-arrays is basically isomorphic to a vector (flatten every array and concatenate them all together). When checking U and V we need to use the in_size of A and C for N and K.

There will need to be a few tree_unflatten's to move between the flattened space (where U and V live) and the pytree input space (where A and C potentially live). This makes the pushthrough operator a bit tricky but should be do-able with a little time.

I suppose my question would be, is there a nice way to wrap this interlink between flattened vector space and pytree space so that implementing this kind of thing will be easier in the future? Does it already exist somewhere outside (or potentially inside) lineax?